File: December - 04/target_subarray_finder.R
input <- scan("stdin", quiet = TRUE)
N <- input[1]
K <- input[2]
arr <- input[3:length(input)]

prefix <- 0
map <- new.env(hash = TRUE)

assign("0", 0, envir = map)

start <- -1
end <- -1

for(i in 1:N){
  prefix <- prefix + arr[i]
  key <- as.character(prefix - K)
  if(exists(key, envir = map)){
    start <- get(key, envir = map)
    end <- i - 1
    break
  }
  assign(as.character(prefix), i, envir = map)
}

cat(start, end)

File: December - 04/Target subarray finder.py
def subarray_sum_indices(arr, K):
    prefix_map = {}
    current_sum = 0
    for i in range(len(arr)):
        current_sum += arr[i]
        if current_sum == K:
            return 0, i
        if current_sum - K in prefix_map:
            return prefix_map[current_sum - K] + 1, i
        if current_sum not in prefix_map:
            prefix_map[current_sum] = i
    return -1, -1
N, K = map(int, input().split())
arr = list(map(int, input().split()))
start, end = subarray_sum_indices(arr, K)
print(start, end)
